posts/simprocs-intro.md

```diff
@@ -0,0 +1,550 @@
+---
+author: 'Yaël Dillies, Paul Lezeau'
+category: 'Metaprogramming'
+date: 2025-03-29 12:33:00 UTC+00:00
+description: 'Brief introduction to simprocs and what they are useful for'
+has_math: true
+link: ''
+slug: what-is-a-simproc
+tags: 'simp, simproc, meta'
+title: What is a simproc?
+type: text
+---
+
+Lean core recently added support for custom simplification procedures, aka *simprocs*.
+This blog post aims to explain what a simproc is, what kind of problems can be solved with simprocs, and what tools we have to write them.
+
+<!-- TEASER_END -->
+
+The first part of this post is a purely informal description of what simprocs are and do.
+The second part is a walkthrough to writing a simple simproc in three different ways.
+
+# What is a simproc
+
+To understand what a simproc is and how it works, we will first explain how `simp` works.
+Then we will give some examples and non-examples of simprocs as well as pointers to analogous concepts in other theorem provers.
+
+## How simp works
+
+On an informal level, `simp` can be seen as recursively traversing the expression `e` to be simplified from the outside-in, each time checking whether the expression matches the left hand side `LHS` of some simp lemma of the form `LHS = RHS`, and replace `LHS` by `RHS` in the expression if it does match.
+
+To see what steps `simp` performs, one can use `set_option trace.Meta.Tactic.simp true`, which prints some of this information in the infoview.
+
+Here is an example of a series of rewrites performed by `simp` on a simple goal:
+```lean
+set_option trace.Meta.Tactic.simp true in
+example : 37 * (Nat.fib 0 + 0) = 0 := by
+  simp
+```
+which yields
+```
+[Meta.Tactic.simp.rewrite] Nat.fib_zero:1000:
+      Nat.fib 0
+    ==>
+      0
+[Meta.Tactic.simp.rewrite] add_zero:1000:
+      0 + 0
+    ==>
+      0
```

Comment:
This trace isn't consistent with the description you give above - if I'm traversing from the outside in, I see _ + 0 before I see Nat.fib 0, so your informal description would simplify first to `37 * Nat.fib 0`

Reply:
That's a good point! I guess this is a downside of being too informal in the first explanation of how simp works roughly. What I think is happening here is that lemmas are also treated as "post" simp procedures, i.e. simp will see `Nat.fib 0 + 0` before it sees `Nat.fib 0`, but it visits subexpressions of `Nat.fib 0 + 0` *before* trying to apply lemmas to `Nat.fib 0 + 0` (or rather what `Nat.fib 0 + 0` has been transformed to after its subexpressions have been visited). 

I'll try to reformulate this section to avoid giving the wrong impression.

Reply:
The trace is gone but I think the underlying phrasing issue remains.

Reply:
I hadn't pushed some of the local changes:) - the phrasing has now changed.